package com.gym.crm.core.security;

import com.gym.crm.core.security.impl.RedisTokenBlacklistService;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.data.redis.core.ValueOperations;
import org.springframework.test.util.ReflectionTestUtils;

import java.util.Date;
import java.util.concurrent.TimeUnit;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class TokenBlacklistServiceTest {
    private static final String TOKEN = "token";
    private static final String BLACKLIST_KEY = "blacklist:";

    @Mock
    private StringRedisTemplate template;

    @Mock
    private JwtService jwtService;

    @InjectMocks
    private RedisTokenBlacklistService tokenBlacklistService;

    @Test
    void isBlacklisted_shouldReturnFalse_whenTokenNotBlacklisted() {
        String hashedToken = ReflectionTestUtils.invokeMethod(tokenBlacklistService, "hashToken", TOKEN);
        when(template.hasKey(BLACKLIST_KEY + hashedToken)).thenReturn(false);

        boolean actual = tokenBlacklistService.isBlacklisted(TOKEN);

        assertThat(actual).isFalse();
        verify(template).hasKey(BLACKLIST_KEY + hashedToken);
    }

    @Test
    void isBlacklisted_shouldReturnTrue_whenTokenBlacklisted() {
        String hashedToken = ReflectionTestUtils.invokeMethod(tokenBlacklistService, "hashToken", TOKEN);
        when(template.hasKey(BLACKLIST_KEY + hashedToken)).thenReturn(true);

        boolean actual = tokenBlacklistService.isBlacklisted(TOKEN);

        assertThat(actual).isTrue();
        verify(template).hasKey(BLACKLIST_KEY + hashedToken);
    }

    @Test
    void blacklist_shouldAddTokenToBlacklist_whenCorrectTime() {
        String hashToken = ReflectionTestUtils.invokeMethod(tokenBlacklistService, "hashToken", TOKEN);
        ValueOperations<String, String> valueOps = mock(ValueOperations.class);
        when(template.opsForValue()).thenReturn(valueOps);
        when(jwtService.extractExpiration(TOKEN)).thenReturn(new Date(System.currentTimeMillis() + 10000));

        tokenBlacklistService.blacklist(TOKEN);

        verify(template).opsForValue();
        verify(valueOps).set(eq(BLACKLIST_KEY + hashToken), eq("true"), anyLong(), eq(TimeUnit.MILLISECONDS));
    }

    @Test
    void blacklist_shouldNotAddToken_whenExpired() {
        when(jwtService.extractExpiration(TOKEN)).thenReturn(new Date(System.currentTimeMillis() - 1000));

        tokenBlacklistService.blacklist(TOKEN);

        verify(template, never()).opsForValue();
    }

    @Test
    void blacklist_and_check_shouldWorkEndToEnd() {
        String token = "token";
        String hashed = ReflectionTestUtils.invokeMethod(tokenBlacklistService, "hashToken", token);
        ValueOperations<String, String> valueOps = mock(ValueOperations.class);

        when(template.opsForValue()).thenReturn(valueOps);
        when(jwtService.extractExpiration(token)).thenReturn(new Date(System.currentTimeMillis() + 10000));
        when(template.hasKey(BLACKLIST_KEY + hashed)).thenReturn(true);

        tokenBlacklistService.blacklist(token);

        boolean result = tokenBlacklistService.isBlacklisted(token);
        assertThat(result).isTrue();
    }
}